src/python/pants/engine/process.py

```diff
@@ -523,7 +523,7 @@ async def find_binary(request: BinaryPathRequest) -> BinaryPaths:
         f"""\
         {shebang}
 
-        set -euox pipefail
+        set -uox pipefail
 
         if command -v which > /dev/null; then
```

Comment:
I'm slow, which branch fails, and why does it fail - `-e` aside. IIUC `command` is a bash builtin and we guard with `|| true` in both branches...

Reply:
Aha - should the whole thing just be a fallible process?

Reply:
Urm, the latter doesn't make sense either given the script. I'm still confused what this fixes and how.

Reply:
I updated the PR description to explain better. This line is fine, the issue is line 529. `which -a` fails and early exits before the `|| true` kicks in.

--

> should the whole thing just be a fallible process?

Maybe? Stu changed it in https://github.com/pantsbuild/pants/pull/10858 w/ this rationale: 

> Additionally, change the "discovery" portion of `BinaryPaths` from `FallibleProcessResult` to `ProcessResult` to fail more quickly in cases where discovery errors, as opposed to succeeding with an empty result.

I'm not really sure what that buys us.

Reply:
Aha, so this had nothing to do with which then since the else branch also used the `|| true` trick incorrectly?

Reply:
From bash man page, NB the "The shell does not exit if the command that fails is part of..." verbage:
```
-e      Exit immediately if a pipeline (which may consist of a single simple command), a list, or a compound command (see SHELL GRAMMAR above), exits with a non-zero status.  The shell does not exit if the command that fails is part of the command list immediately
                      following  a  while or until keyword, part of the test following the if or elif reserved words, part of any command executed in a && or || list except the command following the final && or ||, any command in a pipeline but the last, or if the command's re‐
                      turn value is being inverted with !.
...
```

Reply:
```console
$ bash -c 'set -e; false || true'
$ echo $?
0
$
```

Reply:
Maybe this is an evolution in bash behavior and an old vs new bash thing?

Reply:
Try running this:

```bash
$ set -e
$ command -v fake || true
$ echo $?
# should give back 0
$ command which -a fake || true
# should exit out of your tab
```

If you do just `which -a fake` w/o the `command`, it works as we had intended. I'm not sure why we're using `command` as first arg.

Reply:
Nope:
```console
$ set -e
$ command -v fake || true
$ echo $?
0
$ command which -a fake || true
which: no fake in (/home/jsirois/.rvm/gems/ruby-2.6.0/bin:/home/jsirois/.rvm/gems/ruby-2.6.0@global/bin:/home/jsirois/.rvm/rubies/ruby-2.6.0/bin:/home/jsirois/.cargo/bin:/home/jsirois/.pyenv/shims:/home/jsirois/bin:/home/jsirois/.local/bin:/home/jsirois/.nvm/versions/node/v8.11.3/bin:/home/jsirois/.poetry/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/opt/cuda/bin:/opt/cuda/nsight_compute:/opt/cuda/nsight_systems/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/jsirois/.bash.d/bin:/home/jsirois/go/bin:/home/jsirois/.gem/ruby/2.5.0/bin:/home/jsirois/.rvm/bin:/home/jsirois/.bash.d/bin)
$ echo $?
0
$ 
```

And:
```console
$ bash --version
GNU bash, version 5.1.8(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

Reply:
But thanks for the explanation. Apparently this is old bash behavior (I <3 macOS) and we need to support that. Do you mind adding comments in the code that explain all this or else switching back to a fallible process or anything to squash fuzzy thinking going forward. This is all super non obvious.